mantidproject/mantid
# Documentation system test and standard for interface file names

**Description of work**

Warning: the work for #36161 will have to be updated if this is merged in first. ``MagnetismReflectometryReduction-v1.rst`` will need to be removed from the exceptions list in the test.

This PR is to introduce a system test to check for missing documentation for
 - Algorithms
 - Fit functions
 - Interfaces (accessible through the interfaces menu)

It was also required to enforce some standard naming rules for interface files and docs so that they could be programmatically searched for.

**Purpose of work**
Several algorithms have been added without accompanying docs, leading to 404 pages when clicking the help button. At the moment, the only defence against this is PR reviews, but it's something that can easily be missed.

The new test(s) will do this automatically.

**Summary of work**

_DocumentationTest.py_
 - _AlgorithmDocumentationTest_
   - Uses the algorithm factory to get all the registered algorithms, then searches for a matching doc file within the algorithm docs directory.
   - There is a list of exceptions that's been added to make the test pass, these algorithms need docs writing (at least the bare essentials that we have sphinx commands for).
   - It will fail if there are any missing docs (not in the exception list) or if docs are found but are still in the exceptions list (list needs updating)
  - _FittingDocumentationTest_
    - This test does essentially the same thing as the algorithm one but using the function factory instead. Again there is a list of exceptions which needs to be cleared.
  - _InterfaceDocumentationTest_
  	- To collect the names of the interfaces, I had to refactor some functions out of the main window class and into a new module `gather_interfaces.py` so that they could be used in this test.
  	- This test also follows the same principles but has some rules for searching for the doc files
  	  - Python interface names should have doc files with the same name, but underscores are replaced with spaces
  	  - C++ should ideally have doc files with the same name, but in reality a search is done for `<interface_name>.rst` and `<category> <interface_name>.rst` on both C++ and python interfaces.
  	- The exceptions list has some interfaces which may not need docs (the externally installed ones like `Shiver`). There are also several which call launch `reduction_interface.py` with different combinations of options. Unsure if docs are required for these.
  	- Finally, `Direct Data Analysis` and `Indirect Data Analysis` both open the Inelastic Data Analysis interface, so they don't point to any docs of their own. It feels like maybe the interface menu should change to reflect this.

**Further detail of work**



**To test:**
 - Code review
 - Using [This Build](https://builds.mantidproject.org/job/build_packages_from_branch/639/), check that help docs to all the interfaces which have been touched by this pr open.

Fixes #34168 and fixes #34169





---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
